### PR TITLE
enhance audit log

### PR DESCRIPTION
**Pull request by aiwantaozi**
Issue: https://github.com/rancherlabs/rancher-security/issues/24
improve audit log code.

* For sourceIP we currently parse headers to figure out the ip if not passed in remoteAddress. Instead we should log the remoteAddress + all the headers. So the user can make a choice at which header to look if the remoteAddress is empty, instead of making this decision for him.
* Use byte array type for request and response rather than the interface. 
* Do request and response writing in one entry by maintaining a byte buffer
* Ensure that we do a proper encoding
* Use k8s constant for logging date format (RFC 3339 type)

* Create a separate filter for the logging that would act as a wrapper for all the requests. Do not reuse the auth filter even though it's called on every request(move it back to the original package)
* There is a bug where 401 errors don't get audit logged. That happens because we return on this line w/o calling the logger: https://github.com/rancher/rancher/pull/14157/files#diff-8f73d8fc3a59e92e578239e9e46e4087R71. Once we put a generic filter for all the requests, this issue should be validated

### CODE REVIEW

File: pkg/audit/filter.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	resBody     []byte
+}
+
+func (aw *wrapWriter) WriteHeader(statusCode int) {
+	aw.ResponseWriter.WriteHeader(statusCode)
+	aw.statusCode = statusCode
+}
+
+func (aw *wrapWriter) Write(body []byte) (int, error) {
+	aw.resBody = body
```
Comment (ibuildthecloud) [inline pkg/audit/filter.go]: you need to append here, not assign

File: pkg/audit/audit.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+	var buffer bytes.Buffer
+	alByte, err := json.Marshal(a.log)
+	if err != nil {
+		return err
+	}
+
+	buffer.Write(bytes.TrimSuffix(alByte, []byte("}")))
+
+	if a.writer.Level >= levelRequest && a.reqBody != nil {
```
Comment (ibuildthecloud) [inline pkg/audit/audit.go]: This shouldn't be required as the []byte fields are already encoded from the Marshal call before.

File: pkg/audit/audit.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+func (a *auditLog) write(reqHeaders, resHeaders http.Header, resCode int, resBody []byte) error {
+	userInfo := &userInfo{
+		Name:  reqHeaders.Get("Impersonate-User"),
+		Group: reqHeaders.Get("Impersonate-Group"),
+	}
+	a.log.User = userInfo
+	a.log.ResponseTimestamp = time.Now().Format(time.RFC3339)
+	a.log.RequestHeader = reqHeaders
+	a.log.ResponseHeader = resHeaders
```
Comment (ibuildthecloud) [inline pkg/audit/audit.go]: Don't log Cookie, Set-Cookie headers

File: pkg/audit/audit.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+func (a *auditLog) write(reqHeaders, resHeaders http.Header, resCode int, resBody []byte) error {
+	userInfo := &userInfo{
+		Name:  reqHeaders.Get("Impersonate-User"),
+		Group: reqHeaders.Get("Impersonate-Group"),
+	}
+	a.log.User = userInfo
+	a.log.ResponseTimestamp = time.Now().Format(time.RFC3339)
+	a.log.RequestHeader = reqHeaders
```
Comment (ibuildthecloud) [inline pkg/audit/audit.go]: Don't log the Authorization or Cookie header

File: pkg/audit/audit.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+			AuditID:          k8stypes.UID(uuid.NewRandom().String()),
+			RequestURI:       req.RequestURI,
+			Verb:             req.Method,
+			SourceIP:         getSourceIPs(req.RemoteAddr),
+			RequestTimestamp: time.Now().Format(time.RFC3339),
+		},
+	}
+
+	contentType := req.Header.Get("Content-Type")
+	if writer.Level >= levelRequest && bodyMethods[req.Method] && contentType == contentTypeJSON {
```
Comment (ibuildthecloud) [inline pkg/audit/audit.go]: Do you really care if it's JSON?

File: pkg/audit/audit.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+	log     *log
+	writer  *LogWriter
+	reqBody []byte
+}
+
+type log struct {
+	AuditID           k8stypes.UID `json:"auditID,omitempty"`
+	RequestURI        string       `json:"requestURI,omitempty"`
+	User              *userInfo    `json:"user,omitempty"`
+	Verb              string       `json:"verb,omitempty"`
```
Comment (ibuildthecloud) [inline pkg/audit/audit.go]: Change to Method

File: pkg/audit/audit.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+	SourceIP          string       `json:"sourceIP,omitempty"`
```
Comment (ibuildthecloud) [inline pkg/audit/audit.go]: Change to RemoteAddr

File: pkg/audit/audit.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+	if a.writer.Level >= levelRequestResponse && resBody != nil {
+		buffer.WriteString(`,"responseBody":`)
+		buffer.Write(bytes.TrimSuffix(resBody, []byte("\n")))
+	}
+	buffer.WriteString("}\n")
+	_, err = a.writer.Output.Write(buffer.Bytes())
+	return err
+}
+
+func getSourceIPs(remoteAddr string) string {
```
Comment (ibuildthecloud) [inline pkg/audit/audit.go]: Why don't you just log RemoteAddr and not read, change it at all.

File: pkg/audit/audit.go (hunk 5 above)
Comment (aiwantaozi) [reply pkg/audit/audit.go]: I afraid other format logs will make the audit log mess, hard to parse in other tools.

File: pkg/audit/audit.go (hunk 2 above)
Comment (aiwantaozi) [reply pkg/audit/audit.go]: Do you mean ``` a.reqBody != nil ```? because in some case like not content-type not json the body will be nil.

File: pkg/audit/audit.go (hunk 3 above)
Comment (aiwantaozi) [reply pkg/audit/audit.go]: Do you think we need the headers? The request head and response header are lists below, seems like they are meaningless to most of user.
request header"
```
"requestHeader": {
        "Accept": [
            "application/json"
        ],
        "Accept-Encoding": [
            "gzip, deflate, br"
        ],
        "Accept-Language": [
            "en,zh-CN;q=0.9,zh;q=0.8"
        ],
        "Connection": [
            "keep-alive"
        ],
        "Content-Type": [
            "application/json"
        ],
        "Impersonate-Group": [
            "system:authenticated"
        ],
        "Impersonate-User": [
            "user-bkszq"
        ],
        "Referer": [
            "https://localhost:8443/c/local/nodes/local:machine-r7nfg"
        ],
        "User-Agent": [
            "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.106 Safari/537.36"
        ],
        "X-Api-Action-Links": [
            "actionLinks"
        ],
        "X-Api-Csrf": [
            "853952e7fc"
        ],
        "X-Api-No-Challenge": [
            "true"
        ]
    },
```

response header:
```
"responseHeader": {
        "Content-Type": [
            "application/json"
        ],
        "Expires": [
            "Wed 24 Feb 1982 18:42:00 GMT"
        ],
        "X-Api-Schemas": [
            "https://localhost:8443/v3/schemas"
        ]
    }
```

**Comment by aiwantaozi** [comment]
Do you think we need the headers? The request head and response header are lists below, seems like they are meaningless to most of user.
request header"
```
"requestHeader": {
        "Accept": [
            "application/json"
        ],
        "Accept-Encoding": [
            "gzip, deflate, br"
        ],
        "Accept-Language": [
            "en,zh-CN;q=0.9,zh;q=0.8"
        ],
        "Connection": [
            "keep-alive"
        ],
        "Content-Type": [
            "application/json"
        ],
        "Impersonate-Group": [
            "system:authenticated"
        ],
        "Impersonate-User": [
            "user-bkszq"
        ],
        "Referer": [
            "https://localhost:8443/c/local/nodes/local:machine-r7nfg"
        ],
        "User-Agent": [
            "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.106 Safari/537.36"
        ],
        "X-Api-Action-Links": [
            "actionLinks"
        ],
        "X-Api-Csrf": [
            "853952e7fc"
        ],
        "X-Api-No-Challenge": [
            "true"
        ]
    },
```

response header:
```
"responseHeader": {
        "Content-Type": [
            "application/json"
        ],
        "Expires": [
            "Wed 24 Feb 1982 18:42:00 GMT"
        ],
        "X-Api-Schemas": [
            "https://localhost:8443/v3/schemas"
        ]
    }
```

**Comment by aiwantaozi** [comment]
@ibuildthecloud updated, could you help review?

**Comment by alena1108** [comment]
Merging after getting an offline green light from @ibuildthecloud
